Npgsql#1078 PGPASS file

rmc00 opened the pull request:
Issue #1078 - Added support for PGPASS file and some tests for the code changes

roji commented on src/Npgsql/PgPassFile.cs:
```diff
@@ -0,0 +1,71 @@
+        /// Retrieves the full expected path to the .PGPASS file
+        /// </summary>
+        /// <returns>Path to the .PGPASS file</returns>
+        public static string GetPgPassFilePath()
```
This function needs to also work on Linux/MacOS, can you add support for this?

roji replied:
You can use `PGUtil.IsWindows` to test whether you're on Windows or not

roji commented on src/Npgsql/PgPassFile.cs:
```diff
@@ -0,0 +1,71 @@
+        {
+            get
+            {
+                var serializedCredentials = File.ReadAllLines(FileName);
```
It's better to use ReadLines here (which returns an IEnumerable)

roji replied:
Also, this whole property can be one simple LINQ expression, using an expression-bodied property.

roji commented on test/Npgsql.Tests/ConnectionTests.cs:
```diff
@@ -40,6 +40,31 @@ namespace Npgsql.Tests
+            var pgpassFile = PgPassFile.GetPgPassFilePath();
+            if (File.Exists(pgpassFile))
+            {
+                File.Delete(pgpassFile);
```
Umm, this deletes the actual pgpass.conf file on the system, even if it existed regardless of unit tests... It's very problematic for unit tests to change the environment in this way.

IMHO for this feature it could be enough to test manually. To make a less invasive unit test you could copy any current pgpass aside and restore it at the end of the tests.

Another problem with the code is that you're hard-coding npgsql_tests as the user and password, but the Npgsql testsuite allows you to change the connection string via an environment variable. You should extract the username and password from `TestBase.ConnectionString`.

roji replied:
Also please move this setup code to the bottom of ConnectionTests.cs and wrap it in a Setup/Teardown region.

rmc00 replied:
I've implemented your feedback for the main functionality, but I'm a little confused about this feedback on the tests. Agree completely about what you're saying with destroying the proper pgpass file... I can't believe I didn't realize it before. Are you suggesting that I should go ahead and delete the tests or make it less invasive? If you still see value in the tests, then I will fix them up. While I was adding support for mac/linux file paths, I added code to respect the PGPASSFILE environment variable. That might be a more suitable way to accomplish this. I can store the current value (if set) for the environment variable, create a new temp file with the mock data (adjusting it to use proper username, password, etc from the base class), and then run temporarily with the environment variable pointing to the temp file. In the teardown, I can just restore or delete the environment variable to put it back the way it was. Please let me know your thoughts about that, and I can fix (or delete) the test classes and get the revised code to you.

roji replied:
@rmc00 using the environment variable is a great idea. Restoring the environment variable to whatever it was before the test started (and deleting the mock pgpass file) is a good idea, this way things things are back to exactly what they were. To be extra-careful, please mark these tests with `[Parallelizable(ParallelScope.None)]`, this way NUnit never runs them in parallel (even though parallelization isn't currently on, but it will be at some point).

roji commented on src/Npgsql/NpgsqlConnection.cs:
```diff
@@ -201,7 +201,25 @@ void OpenInternal()
+                    // no password was provided. attempt to pull the password from the pgpass file.
+                    var pgPassFile = new PgPassFile(PgPassFile.GetPgPassFilePath());
+
+                    var host = Settings.Host ?? PgPassEntry.PGPASS_WILDCARD;
```
I think we always have Host and Database (just like we always have Port), so there's no need to use null coalescing here.

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <summary>
+        /// Hostname parsed from the .pgpass file
+        /// </summary>
+        public string Hostname { get; }
```
Please mark all the members of this class internal rather than public (both methods and properties)

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,109 @@
+
+        #endregion
+
+
```
Extra space

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,109 @@
+        #endregion
+
+        internal const string PGPASS_WILDCARD = "*";
+        /// <summary>
```
Missing space

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <param name="database">Database to check against this entry</param>
+        /// <param name="userName">Username to check against this entry</param>
+        /// <returns>True if the entry is a match. False otherwise.</returns>
+        public bool IsMatch(string hostName, string port, string database, string userName)
```
Convert to expression-bodied method

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,109 @@
+        /// <param name="value1">Value being searched</param>
+        /// <param name="value2">Value from the PGPASS entry</param>
+        /// <returns>True if the values are a match. False otherwise.</returns>
+        bool AreValuesMatched(string value1, string value2)
```
Convert to expression-bodied method

roji commented on src/Npgsql/PgPassFile.cs:
```diff
@@ -0,0 +1,71 @@
+        /// <param name="database">Database to query. Use * to match any. </param>
+        /// <param name="userName">User name to query. Use * to match any. </param>
+        /// <returns>Matching <see cref="PgPassEntry"/> if match was found. Otherwise, returns null.</returns>
+        public PgPassEntry GetFirstMatchingEntry(string hostName = "*", string port = "*", string database="*", string userName = "*")
```
Convert to expression-bodied method

roji commented:
Thanks for the PR, overall things look good but please take a look at my comments. It's especially important to have this work on non-Windows platforms as well, where the pgpass file is located in a different place.

rmc00 commented:
@roji Thanks for your feedback. I think this latest push contains fixes for all the comments you made. Please let me know if any other modifications are needed, but I hope this completes the pgpass feature.

roji commented on src/Npgsql/NpgsqlConnection.cs:
```diff
@@ -201,7 +201,22 @@ void OpenInternal()
+                {
+                    Password = Settings.Password;
+                }
+                else if (Settings.Password == null && !IntegratedSecurity)
```
No need to check `Settings.Password == null`, we already know that's the case.

roji replied:
Also, no need to check `!IntegratedSecurity`, it's OK for the password to be set even if IntegratedSecurity is on. This continues the previous behavior where the Password field is populated regardless of IntegratedSecurity.

Although it may look weird, remember that it is PostgreSQL which determines what kind of authentication is going to take place, depending on how pg_hba.conf is set up. Providing both a password and IntegratedSecurity allows users to say "if the server wants a password use that, if it wants SSPI/GSS use that".

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,103 @@
+        /// <summary>
+        /// Port parsed from the .pgpass file
+        /// </summary>
+        internal string Port { get; }
```
It would be cleaner to have this as an int rather than a string, after all this class is aware of the actual structure of the pgpass file and therefore also of the typing.

roji commented on src/Npgsql/NpgsqlConnection.cs:
```diff
@@ -201,7 +201,22 @@ void OpenInternal()
+                    // no password was provided. attempt to pull the password from the pgpass file.
+                    var pgPassFile = new PgPassFile(PgPassFile.GetPgPassFilePath());
+
+                    var user = Settings.Username ?? PgPassEntry.PGPASS_WILDCARD;
```
It is probably a bit cleaner to simply pass a null to the matching function to indicate "any user". Inside the class PGPASS_WILDCARD can still be used of course.

roji replied:
If you do this, remember to add the [CanBeNull] attribute on the user parameter of the match function.

roji commented on src/Npgsql/PgPassEntry.cs:
```diff
@@ -0,0 +1,103 @@
+            var parts = Regex.Split(serializedEntry, @"(?<!\\):"); // split on any colons that aren't preceded by a \ (\ indicates that the colon is part of the content and not a separator)
+            if (parts == null || parts.Length != 5)
+            {
+                throw new FormatException(".PGPASS entry was not well-formed. Please ensure all non-comment entries are formatted as hostname:port:database:username:password. If colon is included, it must be escaped like \\:.");
```
Hmm, I'm not sure it's a good idea to bomb the connection process because of a malformed pgpass - is this what happens with psql (or other libpq programs)? We should match whatever it is they do.

rmc00 replied:
I did some testing with psql using a malformed pgpass file. It returns a message like this:

psql: FATAL:  password authentication failed for user "npgsql_tests"
password retrieved from file "_mypgpassfile_"

Based on that, I assume you want to leave the exception in place?

roji replied:
Hmm... that seems to indicate psql did manage to extract some sort of password from the file (probably mangled because of the bad format), and that it wasn't the right password.

You can play around with it some more to see if psql ever actually complains about a malformed pgpass file, although I guess it's not that important, i..e it's probably OK to leave the exception in place.

roji commented on src/Npgsql/PgPassFile.cs:
```diff
@@ -0,0 +1,66 @@
+        /// Parses file content and gets all credentials from the file
+        /// </summary>
+        /// <returns><see cref="IEnumerable{PgPassEntry}"/> corresponding to all lines in the .pgpass file</returns>
+        internal IEnumerable<PgPassEntry> Entries => File.ReadLines(FileName).Select(line => line.Trim()).Where(line => line.Any() && line[0] != '#').Select(PgPassEntry.Create);
```
Please put each LINQ function on its own line for better readability

roji commented on src/Npgsql/PgPassFile.cs:
```diff
@@ -0,0 +1,66 @@
+        /// <param name="database">Database to query. Use * to match any. </param>
+        /// <param name="userName">User name to query. Use * to match any. </param>
+        /// <returns>Matching <see cref="PgPassEntry"/> if match was found. Otherwise, returns null.</returns>
+        internal PgPassEntry GetFirstMatchingEntry(string hostName = "*", string port = "*", string database="*", string userName = "*") => Entries.FirstOrDefault(entry => entry.IsMatch(hostName, port, database, userName));
```
Break to two lines

roji commented on src/Npgsql/PgPassFile.cs:
```diff
@@ -0,0 +1,66 @@
+            {
+                return pgpassEnv;
+            }
+            return PGUtil.IsWindows ? Path.Combine(Environment.GetFolderPath(Environment.SpecialFolder.ApplicationData), "postgresql", "pgpass.conf") : "~/.pgpass";
```
Hmm... I'm not sure how kosher (and portable) it is to use ~ when opening a file like this. It's probably much better to simply get the home directory (by looking at the HOME environment variable) and combining that.

Also please split this long ternary expression to three lines.

roji commented on test/Npgsql.Tests/ConnectionTests.cs:
```diff
@@ -979,5 +979,53 @@ public void ExceptionDuringClose()
+        string _pgpassFile = null;
+
+        [OneTimeSetUp]
+        public void CreateTestFile()
```
This looks much better than before, but there's still a problem - the setup/teardown runs for all test in the ConnectionTests suite, possibly affecting them. If we have just one test (that's the case), the setup and teardown code should simply be integrated inside the test (the teardown should be in a finally clause). If we have more than one test, it's better to create a different suite with OneTimeSetUp/TearDown as you did, this way everything is properly isolated. Sorry for not making this clear before.

roji commented on test/Npgsql.Tests/Npgsql.Tests.csproj:
```diff
@@ -105,5 +107,8 @@
     <None Include="project.json" />
   </ItemGroup>
+  <ItemGroup>
+    <Service Include="{82A7F48D-3B50-4B1E-B82E-3ADA8210C358}" />
```
Not sure what this is, should probably be removed.

roji commented on test/Npgsql.Tests/PgPassFileTests.cs:
```diff
@@ -0,0 +1,77 @@
+{
+    [TestFixture]
+    [Parallelizable(ParallelScope.None)]
+    public class PgPassFileTests : TestBase
```
If I understand correctly these are true unit tests - they only check PgPass components and don't actually initiate a connection. It's great to have these, but the testsuite shouldn't inherit from TestBase (which is meant for integration tests which do actual database interaction).

For these tests I'd bypass the environment variable and simply instantiate a PgPassFile directly with your own mock path. This reduces impact on the environment, you can remove the `[Parallelizable(ParallelScope.None)]`, etc.

roji commented on test/Npgsql.Tests/PoolTests.cs:
```diff
@@ -70,7 +70,7 @@ public void MinPoolSizeLargeThanPoolSizeLimit()
         public void MinPoolSize()
         {
-            var connString = new NpgsqlConnectionStringBuilder(ConnectionString) { MinPoolSize = 2 };
+            var connString = new NpgsqlConnectionStringBuilder(ConnectionString) { MinPoolSize = 2, PersistSecurityInfo = true };
```
This probably got in here by mistake.

roji commented:
@rmc00, send more comments on your code, I admit a lot of them are nitpicking/style, I appreciate the work you're putting into this. Things look almost ready.

Also, please squash all your commits into one after fixing, to make the history cleaner.

roji commented:
One more thing I just noticed - your code fails to build on netstandard13 (.NET Core), you're probably missing some dependencies. Build the project with dotnet cli to see this locally. If you don't know about all the new stuff (.NET Core, dotnet cli) let me know and I'll take care of that stuff.

rmc00 commented:
@roji I went through your feedback and implemented it. Thanks very much for taking another look through all the code. I've squashed the commits down and pushed it all as one update now. This version has corrected all the issues you found, and I also got it building with netstandard13. It was missing some dependencies like you said, but I also had to change some of the logic to find the user's home directory. Some of the enums and methods I had been using are not supported on .NET Core, so I switched that up a bit for what seems like the current best method to find the user's home directory.

roji commented:
Merged via 9a799ae23fdee760b7daeb8dc75777205f3122a0, some fixes in f7a75afa940bb8f0c434269f7208ff7ba9f8dff4

@rmc00 I added another commit that cleans up some stuff. There's not much important stuff, except maybe that it seems you were looking for pgpass.conf in the wrong place - on Windows the pgpass file can be found under `%APPDATA%\postgresql\pgpass.conf` as documented [here](https://www.postgresql.org/docs/current/static/libpq-pgpass.html), your code looked for it under `%HOMEDRIVE%%HOMEPATH%\postgresql`.

Aside from that I moved PgPassEntry to be an inner class of PgPassFile, changed to use nulls everywhere to signify wildcards and cleaned up some style. Let me know if you have any questions or objections.

And thanks for submitting this!
